### PR TITLE
Command to view template typecheck block

### PR DESCRIPTION
**Pull request by ayazhafiz**
This patch adds a command to retrieve and display the typecheck block
for a template under the user's active selection (if any), and
highlights the span of the node(s) in the typecheck block that
correspond to the template node under the user's active selection (if
any). The typecheck block is made available via a dedicated text
document provider that queries fresh typecheck block content whenever
the `getTemplateTcb` command is invoked.

Probably not something we want to land soon, but a useful debugging tool
for folks working with TCBs.

See also https://github.com/angular/angular/pull/39974, which provides
the language service implementations needed for this feature.

### CODE REVIEW

**Comment by ayazhafiz** [comment]
cc @kyliau if you are interested in the `TextDocumentProvider` implementation

**Comment by ayazhafiz** [comment]
Blocked on a release of angular containing https://github.com/angular/angular/commit/d482f5cdd3843b9b6782f17f3dbb7af660f400dd

**Comment by kyliau** [comment]
@ayazhafiz `@angular/language-service` v11.2.0-next.0 is out, we can proceed with this PR now :)

**Comment by ayazhafiz** [comment]
@kyliau sounds good, I will clean this up soon.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [inline client/src/commands/getTcb.ts:96]: I could find no way to test the client-side behavior on the VSCode docs. Let me know if you have any ideas; for now I recommend just spinning up a session and see what happens. On my side it looked to be okay, I can add a demo if folks desire.

File: server/src/session.ts
Comment (atscott) [inline server/src/session.ts:146]: Typically, the way this was done in the past would be add a stub implementation of `getTcb` in the VE LS that's just `return undefined`. I would prefer that a bit more since it requires less change when we eventually remove the VE version. 
With that, I think you can change the return type of `getLSAndScriptInfo` to be `NgLanguageService`

File: client/src/commands/getTcb.ts
Comment (atscott) [inline client/src/commands/getTcb.ts:24]: Small nit on naming: my brain reads these as `lineLength` and `offsetLength` but it takes longer because it has to always do the conversion. I would personally prefer the words spelled out but I'll survive if you keep it as-is :)

File: package.json
Comment (atscott) [inline package.json:32]: I wonder if it would be good to provide this command in the context menu of html files: https://stackoverflow.com/a/41400631

**Review by atscott — CHANGES_REQUESTED** [review]

**Comment by atscott** [comment]
This is _really_ cool by the way; thanks for taking this on!

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:30]: it's the character offset. jsdoc added

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:40]: It's a void function, annotated as such.

File: server/src/session.ts
Comment (ayazhafiz) [reply server/src/session.ts:146]: Yeah you're right. We'll update on the angular side.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:24]: I am not sure I follow, but I tried to rename the variables to spell everything out. Can you let me know if it is better?

File: package.json
Comment (ayazhafiz) [reply package.json:32]: neat idea! done.

File: package.json
Comment (atscott) [inline package.json:118]: Because these items map directly to commands that are defined above, does it make sense to keep the "commands" item and the "menus" item next to each other?

**Review by atscott — APPROVED** [review]

File: package.json
Comment (ayazhafiz) [reply package.json]: sure!

File: client/src/commands.ts
Comment (atscott) [inline client/src/commands.ts:78]: One thing I noticed here is that the other commands in this list don't depend on the Angular LS being initialized yet, but this one does. We might want to consider _not_ registering this command right away, but instead keep a separate list of commands that we register once the initialization is complete.

File: package.json
Comment (atscott) [inline package.json:42]: One more thing we might want to consider: we can define our own group here if we wanted all Angular commands to appear together in the context menu.

File: client/src/commands/getTcb.ts
Comment (kyliau) [inline client/src/commands/getTcb.ts:18]: I don't think we need this class. Can't we just convert `lsp.Range` to `vscode.Range` by calling the `protocol2CodeConverter`?

**Review by kyliau — CHANGES_REQUESTED** [review]

File: package.json
Comment (ayazhafiz) [reply package.json:42]: There is a nice listing at https://code.visualstudio.com/api/references/contribution-points#Sorting-of-groups.

File: server/yarn.lock
Comment (ayazhafiz) [reply server/yarn.lock:1]: Hm I don't remember why I had to do this.. will investigate more and follow up.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:113]: The return value of `getTemplateTcb` (namely `executeTemplateTcb`) is registered as a callback on the `getTcb` command. When that command is invoked `executeTemplateTcb` is called and we instantiate a document provider, alongside content for the TCB document.

However, the problem at this point is that VSCode updates documents lazily, and will not attempt to recompute a document's content without a notification that it needs to do so. So if we have already requested a TCB, say `a.typecheck.ts`, and then change something in our template and execute the `getTcb` command again, even though we will have updated the document provider and its content, the VSCode engine will not know to pick up these changes. Hence we associate an event emitter with each fresh document provider; VSCode will subscribe to the emitter, and when a new request for `getTcb` is made we fire the event to instruct VSCode to update the document content.

We could listen to changes in the active document, but that is more work than I want to include in this PR; it is a straightforward addition, but I think for these commits we should just get the command working without extra niceties.

Actually I just thought of a better model; I don't think there is any reason we have to reinstantiate a fresh document provider for each request. Again, I'll follow up after some more digging. Sorry, I have been very busy with other commitments so my responses here may be slow.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:18]: Yes, but the problem is that we don't have `lsp.Range`. We have `ts.TextSpan`, which is a pair of character offsets (start, end) which we need to convert into line/column pairs.

File: common/lsp_extensions.ts
Comment (ayazhafiz) [reply common/lsp_extensions.ts:1]: Sound good 👍

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:73]: That is one reason; another was to support syntax highlighting IIRC (but maybe Im wrong). I will follow up and add a comment.

Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:73]: Nevermind I think I did this just because of the readonly mode as you mentioned, not sure why I was thinking of highlighting. added a comment a link to the issue.

File: client/src/commands/getTcb.ts
Comment (atscott) [reply client/src/commands/getTcb.ts:18]: I'm doing a similar conversion in my work with "find components for template" and I'm now using `tsTextSpanToLspRange` in `session.ts` and then from there, converting it to `vscode.Range` in the command. I think that would do the trick here as well?

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:113]: Okay, so the other option I was thinking is that we can have a single document provider and recompute the TCB closed over that provider (and filename, highlights). So it would look something like

```
function makeExecuteGetTcb(...) {
  let currentHighlights;
  let currentTcbFileName;
  let currentTcbContents;
  function update() {
      tcb = getTcb(fileName);
      currentHighlights = getHighlights(tcp);
      currentTcbFileName = tcb.fileName;
      currentTcbContents = tcb.content;
  }
  class TcbProvider extends vscode.TextDocumentProvider {
    onDidChange: new EventEmitter(),
    provideTextDocumentContent(fileName) {
      return currentTcbContents;
    }
  }
  const provider = new TcbProvider();
  vscode.register(provider);
  return function execute() {
    update();
    provider.onDidChange.fire(currentTcbFileName);
    const doc = vscode.open(currentTcbFileName);
    vscode.addDecorations(doc, currentHighlights);
  }
}
```

I'll let you decide which you like more; in either case it's not the most elegant because the filename that we have to compute is not known before we make the getTcb request.

Let me know if you have any more questions; I hope the last comment was able to clear up some confusion if any.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:18]: We also cannot convert from TextSpan to Range on the server side because to do so we would have to assume that the TCB belongs to a script info; while this is true (IIRC), I don't think that's an assumption that the server should make with regard to how the language service handles TCBs and whether it does/doesn't attach them onto a project.

File: server/yarn.lock
Comment (ayazhafiz) [reply server/yarn.lock]: Fixed

File: client/src/commands/getTcb.ts
Comment (ivanwonder) [reply client/src/commands/getTcb.ts:105]: this is my thought with a single document provider. the only thing for the command `angular.getTemplateTcb` to do is just open the document by Uri. The provider will listen to the HTML change and refresh the tcb document.

```ts
class TcbDocumentContentProvider implements vscode.TextDocumentContentProvider, vscode.Disposable {
  private openFileSet = new Set<string>();

  constructor(private client: lsp.LanguageClient) {
    vscode.workspace.onDidOpenTextDocument(doc => {
      if (doc.uri.scheme === 'ngTcb') {
        this.openFileSet.add(doc.uri.query);
      }
    });
    vscode.workspace.onDidChangeTextDocument(doc => {
      if (this.openFileSet.has(doc.document.uri.path)) {
        const uri = vscode.Uri.parse('ngTcb: ngcheck.ts').with({query: doc.document.uri.path});
        this.onDidChangeEmitter.fire(uri);
      }
    });
    vscode.workspace.onDidCloseTextDocument(doc => {
      if (doc.uri.scheme === 'ngTcb') {
        this.openFileSet.delete(doc.uri.query);
      }
    });
    vscode.workspace.registerTextDocumentContentProvider('ngTcb', this);
  }

  dispose() {
    this.onDidChangeEmitter.dispose();
  }

  onDidChangeEmitter = new vscode.EventEmitter<vscode.Uri>();
  onDidChange = this.onDidChangeEmitter.event;

  provideTextDocumentContent(uri: vscode.Uri, token: vscode.CancellationToken):
      vscode.ProviderResult<string> {
    return this.client
        .sendRequest<GetTcbResponse>('ngTcb', {
          textDocument: lsp.TextDocumentIdentifier.create(uri.query),
          position: 0,
        })
        .then(tcb => tcb?.content);
  }
}
```
```ts
this.context.subscriptions.push(new TcbDocumentContentProvider(this.client));
```
```ts
function makeExecuteGetTcb(){
   let uri = vscode.Uri.parse('ngTcb: ngcheck.ts').with({query: activeEditor.document.uri.path});
    let doc = await vscode.workspace.openTextDocument(uri);  // calls back into the provider
    await vscode.window.showTextDocument(doc, {preview: false});
}
```

File: client/src/commands/getTcb.ts
Comment (kyliau) [reply client/src/commands/getTcb.ts:17]: > we would have to assume that the TCB belongs to a script info ...

This will always be the case. Since the server owns the `ProjectService`, which in turn manages all `ScriptInfo`s, converting TS spans to LSP Range should be done by the server. We do the conversion for other methods too, like definition, references, etc.
In fact, it is a leak of abstraction to send TS spans down to the client, because the client should not be aware of TS at all. It only deals with LSP.

File: client/src/commands/getTcb.ts
Comment (kyliau) [reply client/src/commands/getTcb.ts:104]: yeah, we don't have client-side tests / can't run them on CI.
I think a manual test here is sufficient, since this feature is not user-facing.

File: client/src/commands/getTcb.ts
Comment (kyliau) [reply client/src/commands/getTcb.ts:105]: @ayazhafiz with your implementation in the comment above, does that mean we can only keep track of one tcb at a time? Can a provider manage multiple files at a time? I don't think so, but looking at @ivanwonder's suggestion it seems possible?

File: package.json
Comment (kyliau) [reply package.json:42]: +1 to Angular's own group.

File: package.json
Comment (kyliau) [reply package.json:151]: With your change to `@angular/language-service`, we can delete these deps now.

File: server/package.json
Comment (kyliau) [inline server/package.json:18]: same comment as above, please delete.

**Review by kyliau — COMMENTED** [review]
Thanks for looking into this and providing multiple suggestions.
I think it's pretty much good to go, once a few loose ends are taken care of:
1. Span conversion from TS to LSP should be done on the server
2. Create document provider once and use it throughout the lifetime of the extension.
3. Verify if the provider can manage multiple tcbs at a time.

As always, thank you for being so patient :)

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:105]: No, you can keep track of as many TCBs as you want if you have a single provider. In the example I give it only keeps track of one, but yes you can certainly keep track of several. However, I think "keeping track" of previously opened TCB documents is unnecessary, since (1) anytime someone requests a TCB you almost certainly want to make a request to the language server to get fresh TCB contents (I don't think trying to implement any kind of cache is worth it), and (2) if someone closes a TCB and then wants to reopen it, VSCode will have already cached the document and will keep that document content until the provider fires an event marking that the document should be reloaded.

So in my opinion, the only "important" thing is making sure that the event emitter for changed TCBs is properly exposed. To this end I feel the ways to  implement the provider here are the options between
- Creating a fresh document provider each time "getTcb" is requested, passing the same event emitter for requested file names (namely the file name corresponding to the requested TCB). This the current implementation.
- Making one document provider and closing over it, as done in Ivan's example or mine in the comment above Ivan's.

IMO, keeping track of previous document contents is unnecessary as explained above.

File: client/src/commands/getTcb.ts
Comment (ayazhafiz) [reply client/src/commands/getTcb.ts:17]: Would it not be a leaky abstraction to assume that the typecheck block file is attached to the project? This is an internal detail of how the language service creates TCB files and may change later.

I agree the conversion boundary should happen on the server side, you bring up a great point.

**Comment by ayazhafiz** [comment]
> 2. Create document provider once and use it throughout the lifetime of the extension.
> 3. Verify if the provider can manage multiple tcbs at a time.

Please see my comment above in your/Ivan/I's thread about this topic. I think the idea of managing multiple tabs is a misnomer, but creating a single document provider is trivial.

In any case, I do not have time to work on this this week. Feel free to take over and clean up the PR. Otherwise I can take a look this weekend.

**Comment by kyliau** [comment]
Agree that the provider should manage only one tcb at a time, I think this is the behavior of the "markdown preview provider" in vscode. And we should only create a single provider.
Thanks for the great work @ayazhafiz and thanks @ivanwonder for the suggestion!
I'll clean this up and get it merged before the 11.2.0 release tomorrow.

File: client/src/commands/getTcb.ts
Comment (ivanwonder) [reply client/src/commands/getTcb.ts:105]: > (1) anytime someone requests a TCB you almost certainly want to make a request to the language server to get fresh TCB contents (I don't think trying to implement any kind of cache is worth it),

I think I just record the HTML files for the TCB content and refresh the TCB content when the HTML files are changed. The TCB is managed by the uri, one TCB will be opened for the same uri. I have not tested it, but I think vscode will handle the uri. For example, when the user tries to view the TCB of file `files:/app/app.component.html`, we only need to open uri `ngtcb:/app/app.component.html`. The provider will handle the uri and provide the TCB content by sending the request to the service. and when the HTML changed, I will tell the provider the `ngtcb:/app/app.component.html` needs to be refreshed.

> (2) if someone closes a TCB and then wants to reopen it, VSCode will have already cached the document and will keep that document content until the provider fires an event marking that the document should be reloaded.

In my example, I think this will send a request to the service, because the vscode doesn’t know if the content is fresh, but it is ok to send a request(the language service has a cache for it) or add a cache here.

If I missed some info or misunderstood you, sorry for that. Because I think we can’t avoid listening to the HTML files, maybe you do it every time when open the TCB content. Just like you said it’s an option to achieve the function.

Some situations we need to be considered in the future.

- file deleted or renamed. If we need to close the TCB content or not.
- Sync scroll. Maybe we need to send a request to the service and scroll the TCB.
- file persistence. Maybe it’s not so important. After reload the vscode, the vscode will restore the layout and the opened file except for the TCB content.

So perhaps managed by uri is easier to understand for me.

File: client/src/providers.ts
Comment (atscott) [inline client/src/providers.ts:28]: is this supposed to be `tcbUri` and `tcbContent` or am I not understanding what `tcf` stands for?

**Review by atscott — APPROVED** [review]
👍  LGTM

File: client/src/providers.ts
Comment (kyliau) [reply client/src/providers.ts:28]: renamed to `tcbFile` and `tcbContent`.

**Comment by kyliau** [comment]
I cleaned this up a bit:
- create a new group called `angular` for the editor context menu
- make the separation between TS / LSP / vscode clearer
  - the server works with TS, but always sends the response to client in LSP
  - the client works with vscode, and is responsible for serializing / deserializing LSP
  - the common code should only be aware of LSP types
  - as a result, there are now three `GetTcbResponse` types - TS, LSP, and vscode
  - this allowed us to remove special logic for "source maps", the converter correctly handles all conversions
- create a single text document content provider for `ng` files, and use it for all requests.
- register the "get tcb" command as a text editor command (instead of a generic command), so that it is only available if there's an active editor

**Review by kyliau — APPROVED** [review]

**Pull request merged**
